packages/loopring_v3/contracts/lib/DelayedTransaction.sol

```diff
@@ -0,0 +1,321 @@
+
+        bytes4 functionSelector = transaction.data.bytesToBytes4(0);
+        uint delay = getFunctionDelay(transaction.to, functionSelector);
+        if (delay == 0) {
+            (bool success, bytes memory returnData) = exectuteTransaction(transaction);
+            emit TransactionExecuted(
+                transaction.timestamp,
+                transaction.to,
+                transaction.value,
+                transaction.data
+            );
+            // Return the same data the original transaction would
+            // (this will return the data even though this function doesn't have a return value in solidity)
+            assembly {
+                switch success
+                case 0 { revert(add(returnData, 32), mload(returnData)) }
+                default { return(add(returnData, 32), mload(returnData)) }
+            }
+        } else {
+            pendingTransactions.push(transaction);
+            pendingTransactionMap[transaction.id] = pendingTransactions.length;
+            emit TransactionDelayed(
+                transaction.id,
+                transaction.timestamp,
+                transaction.to,
+                transaction.value,
+                transaction.data,
+                delay
+            );
+            totalNumDelayedTransactions++;
+        }
+    }
+
+    function setFunctionDelay(
+        address to,
+        bytes4  functionSelector,
+        uint    delay
+        )
+        internal
+    {
+        // Check if the function already has a delay
+        uint pos = delayedFunctionMap[to][functionSelector];
+        if (pos > 0) {
+            if (delay > 0) {
+                // Just update the delay
+                delayedFunctions[pos - 1].delay = delay;
+            } else {
+                // Remove the delayed function
+                uint size = delayedFunctions.length;
+                if (pos != size) {
+                    DelayedFunction memory lastOne = delayedFunctions[size - 1];
+                    delayedFunctions[pos - 1] = lastOne;
+                    delayedFunctionMap[lastOne.to][lastOne.functionSelector] = pos;
+                }
+                delayedFunctions.length -= 1;
+                delete delayedFunctionMap[to][functionSelector];
+            }
+        } else if (delay > 0) {
+            // Add the new delayed function
+            DelayedFunction memory delayedFunction = DelayedFunction(
+                to,
+                functionSelector,
+                delay
+            );
+            delayedFunctions.push(delayedFunction);
+            delayedFunctionMap[to][functionSelector] = delayedFunctions.length;
+        }
+    }
+
+    function exectuteTransaction(
+        Transaction memory transaction
+        )
+        internal
+        returns (bool success, bytes memory returnData)
+    {
+        // solium-disable-next-line security/no-call-value
+        (success, returnData) = transaction.to.call.value(transaction.value)(transaction.data);
```

Comment:
should we make sure `to != address(this)`?

Reply:
I actually use this in #877 to delay the setting of the USD values in the owner contract itself. So I would definitely keep this.